src/avro.erl

```diff
@@ -309,6 +310,150 @@ decode(avro_json, JSON, TypeOrName, StoreOrLkup, Hook) ->
 decode(avro_binary, Bin, TypeOrName, StoreOrLkup, Hook) ->
   avro_binary_decoder:decode(Bin, TypeOrName, StoreOrLkup, Hook).
 
+%% @doc Checks whether two schemas are compatible in sense that an event
+%% encoded with the writer schema can be decoded by the reader schema.
+%% Schema compatibility is defined in
+%% https://avro.apache.org/docs/1.8.1/spec.html#Schema+Resolution
+%% @end
+-spec is_compatible(avro_type(), avro_type()) -> boolean().
+is_compatible(Reader, Writer) ->
+  try
+    do_is_compatible(Reader, Writer)
+  catch throw:{not_compatible, RPath, WPath} ->
+      {not_compatible, lists:flatten(RPath), lists:flatten(WPath)}
+  end.
+
+desc(RecordType) when ?IS_RECORD_TYPE(RecordType) ->
+  [{record, avro:get_type_name(RecordType)}];
+desc(ArrayType) when ?IS_ARRAY_TYPE(ArrayType) ->
+  [ItemsType] =  desc(avro_array:get_items_type(ArrayType)),
+  [{avro:get_type_name(ArrayType), ItemsType}];
+desc(MapType) when ?IS_MAP_TYPE(MapType) ->
+  [ItemsType] =  desc(avro_map:get_items_type(MapType)),
+  [{avro:get_type_name(MapType), ItemsType}];
+desc(EnumType) when ?IS_ENUM_TYPE(EnumType) ->
+  [{enum, avro:get_type_name(EnumType)}];
+desc(FixedType) when ?IS_FIXED_TYPE(FixedType) ->
+  [{fixed, avro:get_type_name(FixedType)}];
+desc(UnionType) when ?IS_UNION_TYPE(UnionType) ->
+  [{union, avro:get_type_name(UnionType)}];
+desc(AvroType) ->
+  [avro:get_type_name(AvroType)].
+
+do_is_compatible(Reader, Writer)
```

Comment:
we can pass the type stack down in this function's call-stack to make the code more compact:
e.g.
```
do_is_compatible(Reader, Writer, RPath, WPath)
  when ?IS_RECORD_TYPE(Reader) andalso ?IS_RECORD_TYPE(Writer) ->
  ReaderName = avro:get_type_fullname(Reader),
  WriterName = avro:get_type_fullname(Writer),
  ReaderTypes = avro_record:get_all_field_data(Reader),
  WriterTypes = avro_record:get_all_field_data(Writer),
  ReaderName =:= WriterName orelse
    erlang:throw({type_name_changed, {reader, ReaderName}, {writer, WriterName}),
  lists:all(
    fun({FieldName, FieldType, Default}) ->
        case lists:keysearch(FieldName, 1, WriterTypes) of
          false ->
            Default =/= ?NO_VALUE orelse
              erlang:throw({reader_missing_default_value, [FieldName | RPath]);
          {value, {_, WriterType, _}} ->
            do_is_compatible(FieldType, WriterType,
                             [FieldName | RPath], [FieldName | WPath])
        end
    end, ReaderTypes);
do_is_compatible(...) ->
  ...
```